File: src-tauri/src-python/gera/frontmatter.py
"""YAML frontmatter parsing and serialization for Gera markdown files.

Gera notes and projects use YAML frontmatter delimited by ``---``:

    ---
    event_ids:
      - event-1
    project_ids:
      - project-1
    ---
    # My Note Title

    Body content here.
"""

from __future__ import annotations

import re
from typing import Any

import yaml

_FRONTMATTER_RE = re.compile(r"^---\s*\n(.*?)\n---\s*\n?", re.DOTALL)


def parse_frontmatter(content: str) -> tuple[dict[str, Any], str]:
    """Split YAML frontmatter from the markdown body.

    Args:
        content: Raw file content (may or may not have frontmatter).

    Returns:
        A ``(frontmatter_dict, body)`` tuple.
        If no frontmatter is found, returns ``({}, content)``.
    """
    match = _FRONTMATTER_RE.match(content)
    if not match:
        return {}, content

    yaml_str = match.group(1)
    body = content[match.end() :]

    try:
        frontmatter = yaml.safe_load(yaml_str)
    except yaml.YAMLError:
        # If YAML is malformed, treat the whole file as body
        return {}, content

    if not isinstance(frontmatter, dict):
        return {}, content

    return frontmatter, body


def serialize_frontmatter(data: dict[str, Any], body: str) -> str:
    """Combine a frontmatter dict and markdown body into a complete file string.

    Args:
        data: Frontmatter key/value pairs. If empty, no ``---`` block is emitted.
        body: Markdown body content.

    Returns:
        Complete file content string.
    """
    if not data:
        return body

    yaml_str = yaml.dump(data, default_flow_style=False, sort_keys=False).strip()
    # Ensure exactly one newline between frontmatter and body
    body = body.lstrip("\n")
    return f"---\n{yaml_str}\n---\n\n{body}"

File: src/utils/frontmatter.ts
/**
 * YAML frontmatter parsing utilities for client-side.
 * Matches the Python frontmatter.py logic.
 */

const FRONTMATTER_RE = /^---\s*\n([\s\S]*?)\n---\s*\n?/;

export interface ParsedFrontmatter {
  metadata: {
    event_ids?: string[];
    project_ids?: string[];
    [key: string]: any;
  };
  body: string;
}

/**
 * Parse YAML frontmatter from markdown content.
 * Returns the frontmatter metadata and the body without frontmatter.
 */
export function parseFrontmatter(content: string): ParsedFrontmatter {
  const match = content.match(FRONTMATTER_RE);
  
  if (!match) {
    return {
      metadata: {},
      body: content,
    };
  }

  const yamlStr = match[1];
  const body = content.slice(match[0].length);

  try {
    // Simple YAML parsing for our use case (event_ids, project_ids lists)
    const metadata: any = {};
    let currentKey: string | null = null;
    
    for (const line of yamlStr.split('\n')) {
      const trimmed = line.trim();
      
      if (!trimmed || trimmed.startsWith('#')) continue;
      
      // Key: value or Key:
      if (line[0] !== ' ' && line[0] !== '-') {
        const colonIdx = line.indexOf(':');
        if (colonIdx !== -1) {
          currentKey = line.slice(0, colonIdx).trim();
          const value = line.slice(colonIdx + 1).trim();
          if (value) {
            metadata[currentKey] = value;
            currentKey = null;
          } else {
            metadata[currentKey] = [];
          }
        }
      }
      // List item
      else if (trimmed.startsWith('-') && currentKey) {
        const item = trimmed.slice(1).trim();
        if (item) {
          metadata[currentKey].push(item);
        }
      }
    }

    return {
      metadata,
      body,
    };
  } catch (e) {
    // If parsing fails, treat whole content as body
    return {
      metadata: {},
      body: content,
    };
  }
}
